enhancements/cluster-logging/content-filter.md

```diff
@@ -0,0 +1,199 @@
+
+### Goals
+
+The user can configure the ClusterLogForwarder to drop and prune log records according to their requirements.
+
+### Non-Goals
+
+No complex record transformations, for example adding new fields with computed values.
+More complex filters MAY be added in future, but not as part of this enhancement.
+
+## Proposal
+
+### Workflow Description
+
+**log administrator** is a human responsible for configuring log collection, storage and forwarding in a cluster.
+
+1. The log administrator creates named `filter` sections in a `ClusterLogForwarder`
+2. Pipelines that require filtering have a `filterRef` field with a list of filter names.
+3. The collector edits log streams according to the filters before forwarding.
+
+### API Extensions
+
+A new `filters` section in the `ClusterLogForwarder` allows named filters to be defined.
+This proposal defines two types of filter "prune" and "drop".
+
+#### Prune filters
+
+A "prune" filter removes fields from each record passing through the filter.
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name: ""        # User selects filter name
+    type: prune
+    prune:
+      in: []      # Array of field paths, remove fields in the array.
+      notIn: []   # Array of field paths, remove all fields that are NOT in the array.
+```
+
+##### Examples
+
+``` yaml
+  spec:
+    filters:
+    - name: foo
+      type: prune
+      prune:
+		notIn: [.message, .kubernetes]   # Keep only the message and kubernetes fields.
+        in: [.kubernetes.flat_labels]    # Prune the kubernetes.flat_labels sub-field.
+    pipelines:
+    - name: bar
+		filterRefs: ["foo"]
+```
+
+#### Drop filters
+
+##### API
+
+A drop filter applies a sequence of tests to a log record and drops the record if any test passes.
+Each test contains a sequence of conditions, all conditions must be true for the test to pass.
+
+``` yaml
+spec:
+  filters:
+  - name:             # Provided by the user
+    type: drop
+    drop:
+      - test:
+        - field:      # JSON path to the field
+		  # Requires exactly one of the following conditions.
+          matches:    # regular expression match
+          notMatches: # regular expression does not match
+```
+
+Note:
+- If _all_ conditions in a test are true, the test passes.
+- If _any_ test in the drop filter passes, the record is dropped.
+- If there is an error evaluating a condition (e.g. a missing field), that condition evaluates to false.
+  Evaluation continues as normal.
+
+The drop filter is equivalent to a boolean OR of AND clauses. Any boolean expression can be reduced to this form.
+
+##### Example
+
+``` yaml
+filters:
+  - name: important
+    type: drop
+    drop:
+      - tests:
+		- field: .kubernetes.namespace_name
+		  notMatches: "very-important"  # Keep everything from this namespace.
+		- field: .level # Keep important levels
+		  matches: "warning|error|critical"
+```
```

Comment:
```suggestion
``` yaml
filters:
  - name: important
    type: drop
    drop:
      - tests:
         - field: .kubernetes.namespace_name
           notMatches: "very-important"  # Keep everything from this namespace.
         - field: .level # Keep important levels
           matches: "warning|error|critical"
```
```